Implement principal renaming in LDAP

SaharahSarah opened the pull request:
The previous way of renaming principals by adding a new entry and deleting the old one did not work in LDAP due to issues such as requiring renaming the DN in standalone principals, and modifying the DN in place for non-standalone principals. A new DAL function for renaming principals is added so that the plugins can properly rename the principal for their particular database. A function is also added to libkdb_ldap that implements the DAL rename_principal function that properly renames the DN and attributes when necessary.

[ticket: 8065](https://krbdev.mit.edu/rt/Ticket/Display.html?id=8065)

greghudson commented:
We'll have to think about how this new kdb5 function interacts with iprop.  At a minimum some internal documentation might be required.

Assuming we don't modify the iprop protocol (which is certainly easiest), then we don't need a krb5int_rename_principal_no_log() function, and I would prefer that its contents be folded into krb5_db_rename_principal().

Also, I think the second call to log_princ_update() should have FALSE as the third argument.  It would be good to add a rename test to t_iprop.py, probably around line 344 after the delete test and before the forced full resync.

(I haven't looked over the third commit yet.  The first commit seemed okay with the exception that some comments could be improved.)

tlyu commented:
krb5_dbe_specialize_salt() appears to unconditionally convert all key data entries to special salt. It should probably only do so for normal and norealm salt types. I'm not sure about the onlyrealm salt type; we probably don't expect to change realm names of a principal during a rename operation, but I could be missing something (maybe cross-realm principals?).

greghudson commented:
The current rename behavior converts all key data to the special salt type.  If we want to change that behavior I would make it a separate commit, but I don't see a need to change the behavior.

SaharahSarah commented:
Iprop is fixed. krb5_db_rename_principal fails if the plugin doesn't use krb5_db_def_rename_principal as the rename function.

greghudson commented:
I think krb5_dbe_specialize_salt() should copy the computed salt into a memory region allocated with krb5_db_alloc(), since that memory will later be freed inside the DB module.  (The existing salt transformation code in svr_principal.c is incorrect for not doing so.)  See kdb_cpw.c for examples of this being done in krb5_dbe_* functions.

The comment for krb5_dbe_specialize_salt() could be improved.  I suggest "Modify the key data of entry to explicitly store salt values using the KRB5_KDB_SALTTYPE_SPECIAL salt type."

greghudson commented:
In the second commit:

* We try to avoid using the word "plugin" as a noun when writing formally.  So in the commit message and in comments, refer to "KDB modules" rather than "plugins".

* Use imperatives rather than passive voice in the commit message: "Add a new DAL function to support renames in all KDB modules, with a default implementation that performs the old functionality..."

* In the rename_principal comment, use two spaces after periods.  "the call will fail if iprop is enabled" should say "the rename will fail if iprop logging is enabled."  (The distinction is that iprop logging happens on the master, whereas "iprop is enabled" could apply to a master or a slave.)

* log_princ_update() is unused and can be removed from the commit.

* The default function for renaming should be set in kdb_setup_opt_functions(), and the DB2 function table should just have NULL in this spot.  (This design dates from when we were avoiding circular dependencies between kdb5 and KDB modules, so it's a little weird, but we should follow it for consistency.)  The default implementation should also be in kdb_default.c rather than kdb5.c.  I don't think there should be any need to include it in the library export list.  (Some, but not all, of the current default functions are in the export list, but I think that's an oversight as they shouldn't need to be there.)

* In  krb5_db_rename_principal() where we check for custom rename handlers on an iprop master, there is no need to check for v->rename_principal == NULL, but there should be a comment explaining why we have this restriction.  The commit message should also briefly mention this limitation.

* In krb5_db_def_rename_principal():
** Why begin by clearing the error message?  krb5_clear_error_message() should generally only be called when an error code was received from a subsidiary function which might have set an extended message, but the error is being handled internally or ignored and not passed up the chain.
** The comment "Transform salts as necessary" isn't very communicative.  "Store salt values explicitly so that they don't depend on the principal name." would be better.
** Where you replace the principal name, it would be safer to set kdb->princ = NULL right after freeing it, instead of when you see an error from krb5_copy_principal().  (I don't understand your comment "so freeing the dbe doesn't lose target data"; what you are avoiding is a double free of the original principal name.)
** There's another violation of malloc hygiene here; krb5_copy_principal() allocates using the libkrb5 allocator but the name is freed inside the KDB module.  We could finesse this issue by setting kdb->princ just before calling krb5_db_put_principal() and then resetting it back to its old value immediately after.

* Adding a function to kdb_vftabl is an incompatible change, so KRB5_KDB_DAL_MAJOR_VERSION needs to be bumped in kdb.h.

greghudson commented on src/lib/kdb/kdb5.c:
```diff
@@ -2260,6 +2260,46 @@ krb5_dbe_compute_salt(krb5_context context, const krb5_key_data *key,
+
+        entry->key_data[i].key_data_type[1] = KRB5_KDB_SALTTYPE_SPECIAL;
+        krb5_db_free(context, entry->key_data[i].key_data_contents[1]);
+        entry->key_data[i].key_data_contents[1] = (krb5_octet *)data;
```
You can avoid this cast by making data of type uint8_t * (or krb5_octet *, but we are phasing out krb5-specific type names) instead of char *.

greghudson commented on src/lib/kdb/kdb5.c:
```diff
@@ -2260,6 +2260,46 @@ krb5_dbe_compute_salt(krb5_context context, const krb5_key_data *key,
+            goto cleanup;
+        }
+        memcpy(data, salt->data, salt->length);
+        free(salt->data);
```
It would make more sense to free salt using krb5_free_data() at line 2295.

greghudson commented on src/plugins/kdb/ldap/libkdb_ldap/ldap_misc.c:
```diff
@@ -1069,6 +1117,20 @@ krb5_add_int_mem_ldap_mod(LDAPMod ***list, char *attribute, int op, int value)
+    }
+
+    return ret;
+}
```
The DN argument should be named "dn", not "DN".  This function can be shortened and made more correct (using the correct type for the return valid of ldap_modify_ext_s()) as follows:

    int st;

    st = ldap_modify_ext_s(ld, dn, mods, NULL, NULL);
    return (st == LDAP_SUCCESS) ? 0 : set_ldap_error(context, st, op);

greghudson commented:
In the final commit:

* Please keep variable names to lowercase; "DN" and "RDN" components should be "dn" and "rdn".

* is_standalone_principal(), unparse_principal_name(), and rename_principal_dn() need comments above their definitions.  I would also suggest naming the third function rename_principal_rdn().

* For a standalone principal, you are currently renaming the entry with deleteoldrdn=0 and then removing the old krbPrincipalName value.  Why not just rename it with deleteoldrdn=1?

* Tests should try very hard to avoid sleeping.  I think we can test the last modified data change without sleeping by looking at the principal name rather than the timestamp.  kadmin.local has a -p flag which can be used to affect the principal used for last modified data; it can be set to anything (even a principal name which doesn't exist) because kadmin.local authenticates locally.

greghudson commented on src/lib/kdb/kdb_default.c:
```diff
@@ -538,3 +538,52 @@ krb5_def_fetch_mkey_list(krb5_context        context,
+    if (ret)
+        goto cleanup;
+
+    ret = krb5_copy_principal(kcontext, target, &new);
```
We don't need to make this copy; we can just assign kdb->princ = target temporarily.

SaharahSarah commented:
#### https://github.com/krb5/krb5/pull/446#issuecomment-218493289

> I think krb5_dbe_specialize_salt() should copy the computed salt into a memory region allocated with krb5_db_alloc(), since that memory will later be freed inside the DB module. (The existing salt transformation code in svr_principal.c is incorrect for not doing so.) See kdb_cpw.c for examples of this being done in krb5_dbe_* functions.

Copied the salt into a memory region allocated with krb5_db_alloc.

> The comment for krb5_dbe_specialize_salt() could be improved. I suggest "Modify the key data of entry to explicitly store salt values using the KRB5_KDB_SALTTYPE_SPECIAL salt type."

Modified the comment with the suggestion.

#### https://github.com/krb5/krb5/pull/446#issuecomment-218506644

> In the second commit:
>
> * We try to avoid using the word "plugin" as a noun when writing formally.  So in the commit message and in comments, refer to "KDB modules" rather than "plugins".

Replaced "plugin" with "KDB Module".

> * Use imperatives rather than passive voice in the commit message: "Add a new DAL function to support renames in all KDB modules, with a default implementation that performs the old functionality..."

Reworded the commit with imperatives instead of passive voice..

> * In the rename_principal comment, use two spaces after periods.  "the call will fail if iprop is enabled" should say "the rename will fail if iprop logging is enabled."  (The distinction is that iprop logging happens on the master, whereas "iprop is enabled" could apply to a master or a slave.)

Added two spaces and reworded the comment.

> * log_princ_update() is unused and can be removed from the commit.

Removed log_princ_update()

> * The default function for renaming should be set in kdb_setup_opt_functions(), and the DB2 function table should just have NULL in this spot.  (This design dates from when we were avoiding circular dependencies between kdb5 and KDB modules, so it's a little weird, but we should follow it for consistency.)  The default implementation should also be in kdb_default.c rather than kdb5.c.  I don't think there should be any need to include it in the library export list.  (Some, but not all, of the current default functions are in the export list, but I think that's an oversight as they shouldn't need to be there.)

Set the default function for reaname in `kdb_setup_opt_functions()` instead of individually in the modules. If a module doesn't want to support renaming then it will have to manually add a function that returns KRB5_PLUGIN_OP_NOTSUPP.

`krb5_db_def_rename_principal()` was already in kdb_default.c and never in kdb5.c, so nothing was to be done with this.

Removed `krb5_db_def_rename_principal()` from export list.

> * In  krb5_db_rename_principal() where we check for custom rename handlers on an iprop master, there is no need to check for v->rename_principal == NULL, but there should be a comment explaining why we have this restriction.  The commit message should also briefly mention this limitation.

Comment explaining the restriction added, and commit message modified to mention the limitation.

> * In krb5_db_def_rename_principal():
>   * Why begin by clearing the error message?  krb5_clear_error_message() should generally only be called when an error code was received from a subsidiary function which might have set an extended message, but the error is being handled internally or ignored and not passed up the chain.

A lot of the code in `krb5_db_def_rename_principal()` was copied from `kadm5_rename_principal()`, which did this already. I'm not sure why it did, but it has been removed from `krb5_db_def_rename_principal()`
 
>   * The comment "Transform salts as necessary" isn't very communicative.  "Store salt values explicitly so that they don't depend on the principal name." would be better.

The comment was copied from `kadm5_rename_principal()`. Modified the comment to be more descriptive.

>   * Where you replace the principal name, it would be safer to set kdb->princ = NULL right after freeing it, instead of when you see an error from krb5_copy_principal().  (I don't understand your comment "so freeing the dbe doesn't lose target data"; what you are avoiding is a double free of the original principal name.)

This was copied from `kadm5_rename_principal()`. I have absolutely no idea why they were doing that. It seems that the original author thought that `krb5_copy_principal()` wasn't a deep copy.

>   * There's another violation of malloc hygiene here; krb5_copy_principal() allocates using the libkrb5 allocator but the name is freed inside the KDB module.  We could finesse this issue by setting kdb->princ just before calling krb5_db_put_principal() and then resetting it back to its old value immediately after.

The entry principal now is set directly to the target principal before calling, then set back to the old principal right after the call.

> * Adding a function to kdb_vftabl is an incompatible change, so KRB5_KDB_DAL_MAJOR_VERSION needs to be bumped in kdb.h.

Incremented `KRB5_KDB_DAL_MAJOR_VERSION`

#### kdb5.c: 2291

> You can avoid this cast by making data of type uint8_t * (or krb5_octet *, but we are phasing out krb5-specific type names) instead of char *.

Changed the type of `data` to `uint8_t *`

#### kdb5.c: 2287

> It would make more sense to free salt using krb5_free_data() at line 2295.

The salt is now freed using `krb5_free_data()`

#### ldap_misc.c: `krb5_ldap_modify_ext()`

> The DN argument should be named "dn", not "DN".  This function can be shortened and made more correct (using the correct type for the return valid of ldap_modify_ext_s()) as follows:
>
>     int st;
>
>     st = ldap_modify_ext_s(ld, dn, mods, NULL, NULL);
>     return (st == LDAP_SUCCESS) ? 0 : set_ldap_error(context, st, op);

`DN` set to `dn`. `ldap_modify_ext_s()` shortened using suggested changes.

#### https://github.com/krb5/krb5/pull/446#issuecomment-218804442

> In the final commit:

> * Please keep variable names to lowercase; "DN" and "RDN" components should be "dn" and "rdn".

This was done this way based off the surrounding code style. Set all the uppercase variable names to lowercase.

> * is_standalone_principal(), unparse_principal_name(), and rename_principal_dn() need comments above their definitions.  I would also suggest naming the third function rename_principal_rdn().

Added comments above the defintions, and renamed the third function.

> * For a standalone principal, you are currently renaming the entry with deleteoldrdn=0 and then removing the old krbPrincipalName value.  Why not just rename it with deleteoldrdn=1?

I personally am testing this on 389ds. The call failes if deleteoldrdn=1. Not sure if OpenLDAP fails as well but since at least one popular LDAP server fails then it should stay the current way of deleting the old rdn later.

> * Tests should try very hard to avoid sleeping.  I think we can test the last modified data change without sleeping by looking at the principal name rather than the timestamp.  kadmin.local has a -p flag which can be used to affect the principal used for last modified data; it can be set to anything (even a principal name which doesn't exist) because kadmin.local authenticates locally.

The test now uses the -p flag to kadmin.local instead of sleeping. I didn't know if sleep was appropriate, but earlier in the file there's a point where sleep is used as well.

greghudson commented:
> I personally am testing this on 389ds. The call failes if deleteoldrdn=1. Not sure if OpenLDAP fails as well but since at least one popular LDAP server fails then it should stay the current way of deleting the old rdn later.

We should definitely try to work on 389ds, but I can't think of a reason why deleteoldrn=1 should cause that operation to fail, and I think we should investigate further.  Can you reproduce this error and say what the reported error is?

SaharahSarah commented:
The error occurs in both the code and if I run it from the command line.

```
dn: krbprincipalname=test@TEST.KRB5.SARAH,cn=TEST.KRB5.SARAH,cn=krbcontainer,dc=krb5,dc=sarah
changetype: modrdn
newrdn: krbprincipalname=test1@TEST.KRB5.SARAH
deleteoldrdn: 1

modifying rdn of entry "krbprincipalname=test@TEST.KRB5.SARAH,cn=TEST.KRB5.SARAH,cn=krbcontainer,dc=krb5,dc=sarah"
ldap_rename: Operations error (1)
```

greghudson commented:
I can reproduce this problem, and I think it's a bug in 389ds.  I will try to report it upstream, but we will want to work with broken versions of 389ds at least for the moment.  Please add a comment to the function explaining why we're modifying the RDN in an awkward way; it's very important to leave a clear trail when working around other people's bugs in our code.

greghudson commented:
I filed this upstream bug report: https://fedorahosted.org/389/ticket/48848

SaharahSarah commented:
> I can reproduce this problem, and I think it's a bug in 389ds. I will try to report it upstream, but we will want to work with broken versions of 389ds at least for the moment. Please add a comment to the function explaining why we're modifying the RDN in an awkward way; it's very important to leave a clear trail when working around other people's bugs in our code.

The comment was already added. Thank you for filing the bug report.

greghudson commented:
Pushed to master as 09cdb3079508f9d3fcc107a8ee8e7537f70c9d37 c38838be956ce72fcd7142f14bc374dc13dd8bb2 2ac75e548afadde4f87f20fcc1ee1472cdac3fed.  I noticed two naming issues in the third commit and want to call attention to them:

* replace_rdn() and rename_principal_rdn() had parameters named "newdn" and a local variable named "newdn_out".  This is reversed from our usual practices for new code; we use the suffix _out on output parameter names, not for local variables which are used to temporarily hold results.  In both of those functions, I changed the parameter name to use the _out suffix and removed the local variable (which we don't need if the last operation in the function can set the output parameter).

* In unparse_principal_name(), the variable "res" held the krb5_error_code result of subsidiary calls.  It's okay to go with the prevailing name in the file being modified, or with "ret" which is our usual practice for new code, but try to avoid other variants.  I changed the name of res to "st", which is the prevailing name in the file for this purpose, and renamed the user parameter to user_out.

There were also a lot of multi-sentence comments with only one space after the period; I fixed those.
